cpp/src/arrow/dataset/scanner_test.cc

```diff
@@ -1074,6 +1074,24 @@ TEST_P(TestScanner, ProjectedScanNested) {
   AssertScanBatchesUnorderedEqualRepetitionsOf(MakeScanner(batch_in), batch_out);
 }
 
+TEST_P(TestScanner, ProjectedScanNestedFromNames) {
+  SetSchema({
+      field("struct", struct_({field("i32", int32()), field("f64", float64())})),
+      field("nested", struct_({field("left", int32()),
+                               field("right", struct_({field("i32", int32()),
+                                                       field("f64", float64())}))})),
+  });
+  ASSERT_OK_AND_ASSIGN(auto descr,
+                       ProjectionDescr::FromNames({"struct.i32", "nested.right.f64"},
+                                                  *options_->dataset_schema))
+  SetProjection(options_.get(), std::move(descr));
+  auto batch_in = ConstantArrayGenerator::Zeroes(GetParam().items_per_batch, schema_);
+  auto batch_out = ConstantArrayGenerator::Zeroes(
+      GetParam().items_per_batch,
+      schema({field("struct.i32", int32()), field("nested.right.f64", float64())}));
```

Comment:
I would have expected the field names here to be "i32" and "f64"? (consistent with the test above)

Reply:
Ya...we do the field name renaming on python, and then in C++ it's left with whatever the dotted path was. Do you think we should move the field renaming into C++ instead?

Reply:
But the name gets set in `ProjectionDescr::FromNames`, no? (this PR is currently not touching any python, except for a test)

Reply:
Ah of course, mind was reverting to the first iteration of this PR. :sweat_smile:

Reply:
But so how did the tests pass before the latest change as well? It's not actually verifying the expected schema?

Reply:
Peeking around at it now, it doesn't seem so. More checking values; will look more and maybe see if I could get it to actually compare them.

Reply:
Found and fixed in https://github.com/apache/arrow/pull/14326/commits/ae5b01d2ce7c67fba5fc17028e6230fb550d2535, verified it'll fail with different field names. All other tests continued to pass locally, so didn't bother to propagate the `check_metadata` parameter up further from `AssertBatchEquals`.